ExtraCredit/ExtraCredit/src/Main.cpp

```diff
@@ -0,0 +1,71 @@
+#include <iostream>
+#include <sstream>
+#include <memory>
+#include "Shape.h"
+#include "Triangle.h"
+#include "TestHarness.h"
+#include "ProjectCode_API.h"
+#include"DerivedShapes.h"
+#include <windows.h> 
+#include <stdio.h> 
+//#include "../../DIIA/DerivedShapes/DerivedShapes.h"
+//#include "../../DIIB/DerivedShapes/DerivedShapes.h"
+
+//typedef int(__cdecl* MYPROC)(LPCWSTR);
+
+int main()
+{
+
+//std::string str = "D:\\UW\\ExtraCredit\\ExtraCredit\\DIIB\\DerivedShapes\\x64\\Debug\\DerivedShapes.dll";
+//std::wstring temp = std::wstring(str.begin(), str.end());
+//LPCWSTR path1 = temp.c_str();
+
+//std::string str2 = "D:\\UW\\ExtraCredit\\ExtraCredit\\DIIB\\DerivedShapes\\x64\\Debug\\DerivedShapes.dll";
+//std::wstring temp2 = std::wstring(str.begin(), str.end());
+//LPCWSTR path2 = temp.c_str();
+
+//HINSTANCE hDll1 = LoadLibrary(path1);
+//HINSTANCE hDll2 = LoadLibrary(path2);
+
+
+//SetDllDirectory(path1);
+
+	//TestResult tr;
+	//TestRegistry::runAllTests(tr);
+
+	//char test; 
+	//std::cin >> test; 
+
+	//HINSTANCE dsLib;
+	//MYPROC ProcAdd;
+	//BOOL fFreeResult, fRunTimeLinkSuccess = FALSE;
+
+	//// Get a handle to the DLL module.
+
+	//dsLib = LoadLibrary(TEXT("DerivedShape.dll"));
+
+	//// If the handle is valid, try to get the function address.
+
+	//if (dsLib != NULL)
+	//{
+	//	ProcAdd = (MYPROC)GetProcAddress(dsLib, "DerivedShape");
+
+	//	// If the function address is valid, call the function.
+
+	//	if (NULL != ProcAdd)
+	//	{
+	//		fRunTimeLinkSuccess = TRUE;
+	//		(ProcAdd)(L"Message sent to the DLL function\n");
+	//	}
+	//	// Free the DLL module.
+
+	//	fFreeResult = FreeLibrary(dsLib);
+	//}
+
+	//// If unable to call the DLL function, use an alternative.
+	//if (!fRunTimeLinkSuccess)
+	//	printf("Message printed from executable\n");
```

Comment:
With the following changes to your code, I was able to successfully create a red `Rectangles` object.

Main thing seemed to be that your MYPROC (maybe borrowed from some example?) was defining a function type that takes a string parameter. The function you're looking for (`generateRedRectangle`) doesn't take a parameter. And, you needed to have the GetProcAddress call actually look for your particular function.

Plus, do something smart to switch between loading the red and blue DLLs at run time...

```suggestion
#define RUNTIME_LOAD
int main()
{
#ifdef RUNTIME_LOAD
	typedef Rectangles* (__cdecl* GenerateRectangleFuncType)();

	//do something smarter to get path / switch between red/blue paths!
	std::wstring path1{ L"A:\\UW\\510\\ExtraCredit\\JonPhan\\ExtraCredit\\DLL_A\\x64\\Debug" };
    SetDllDirectory(path1.c_str());

	HINSTANCE dsLib;
	BOOL fFreeResult, fRunTimeLinkSuccess = FALSE;

	// Get a handle to the DLL module.

	dsLib = LoadLibrary(TEXT("DerivedShapes.dll"));

	// If the handle is valid, try to get the function address.

	if (dsLib != NULL)
	{
		auto pfCreateRectangle = reinterpret_cast<GenerateRectangleFuncType>(GetProcAddress(dsLib, "generateRedRectangle"));
		if (pfCreateRectangle == NULL)
		{
			std::cout << "pointer to generateRedRectangle cound NOT be obtained" << std::endl;
			return -1;
		}

		// If the function address is valid, call the function.

		if (NULL != pfCreateRectangle)
		{
			fRunTimeLinkSuccess = TRUE;
			Rectangles* pRect{ pfCreateRectangle() };
			std::cout << "Success! " << pRect;
		}
		// Free the DLL module.

		fFreeResult = FreeLibrary(dsLib);
	}

	// If unable to call the DLL function, use an alternative.
	if (!fRunTimeLinkSuccess)
		printf("Message printed from executable\n");


#else
///... the unit-test code
#endif
```